[Bug]: ACP turns fail with "Internal error" after upgrade to 2026.4.25 — gateway WS handshake rejected

### Bug type

Regression (worked before, now fails)

### Beta release blocker

No

### Summary

After upgrading from OpenClaw 2026.4.24 to 2026.4.25, all ACP sessions_spawn calls fail instantly (~2s, 0 tokens) with AcpRuntimeError: Internal error / ACP_TURN_FAILED; gateway logs show a WS client sending sessions.list before the required connect frame, and the bundled acpx v0.6.1 backend registers but never reaches "ready."

### Steps to reproduce

1. Upgrade OpenClaw from 2026.4.24 to 2026.4.25 (`npm install -g openclaw@2026.4.25`).
2. Restart gateway (`openclaw gateway restart`).
3. Spawn any ACP session: `sessions_spawn({ task: "Say hello", runtime: "acp", agentId: "claude", mode: "run" })`.
4. Observe instant failure (~2s, 0 tokens): `AcpRuntimeError: Internal error`.
5. Check gateway log: `gateway connect failed: Error: gateway closed (1008): invalid handshake: first request must be connect`.

### Expected behavior

In 2026.4.24 (bundled acpx v0.5.3), ACP sessions spawned successfully. The acpx backend showed "registered" then "ready" within ~0.2s, and ACP turns completed normally with token output.

### Actual behavior

In 2026.4.25 (bundled acpx v0.6.1), the acpx backend shows "registered" but never "ready." Every ACP turn fails in ~2s with 0 tokens processed. Gateway log shows:

[gateway/ws] cause: invalid-handshake, lastFrameMethod: sessions.list, handshakeError: invalid handshake: first request must be connect
[ERROR] gateway connect failed: Error: gateway closed (1008): invalid handshake: first request must be connect
[gateway] acp startup identity reconcile (renderer=v1): checked=22 resolved=0 failed=22

Reproduced across 6+ gateway restarts, fresh auth token, cleaned session state, and bumped maxConcurrentSessions to 100. Non-ACP subagent spawns work fine.

### OpenClaw version

2026.4.25 (aa36ee6)

### Operating system

macOS 15.4 (Darwin 25.4.0, arm64, Apple M5 Max)

### Install method

npm global (nvm, Node v22.22.2)

### Model

anthropic/claude-sonnet-4-6 (for the parent agent); ACP agents: claude-agent-acp v0.26.0, codex-acp

### Provider / routing chain

openclaw -> anthropic (via claude-cli auth profile)

### Additional provider/model setup details

- Auth profile: anthropic:claude-cli (OAuth, re-authed fresh during debugging — did not resolve the issue)
- OPENCLAW_ACPX_RUNTIME_STARTUP_PROBE: not set
- plugins.entries.acpx.config.permissionMode: approve-all
- plugins.entries.acpx.config.agents.claude.command: claude-agent-acp
- plugins.entries.acpx.config.agents.codex.command: codex-acp
- acp.maxConcurrentSessions: 100
- acp.runtime.ttlMinutes: 45
- Plugin runtime deps path: ~/.openclaw/plugin-runtime-deps/openclaw-2026.4.25-1674699698bd/ (acpx v0.6.1)
- Previous working: ~/.openclaw/plugin-runtime-deps/openclaw-2026.4.24-1674699698bd/ (acpx v0.5.3)

### Logs, screenshots, and evidence

```shell
Gateway log (relevant lines across multiple boots on 2026-04-27):


2026-04-26T16:33:02 [plugins] embedded acpx runtime backend ready          ← last "ready" (2026.4.24)

2026-04-27T13:23:11 [plugins] embedded acpx runtime backend registered     ← first boot on 2026.4.25, never "ready"
2026-04-27T13:48:01 [plugins] embedded acpx runtime backend registered     ← second boot, never "ready"
2026-04-27T14:43:51 [plugins] embedded acpx runtime backend registered     ← third boot, never "ready"

2026-04-27T14:47:09 [ws] res ✗ agent errorCode=UNAVAILABLE errorMessage=AcpRuntimeError: Internal error code=ACP_TURN_FAILED
2026-04-27T14:48:42 [ws] res ✗ agent errorCode=UNAVAILABLE errorMessage=AcpRuntimeError: Internal error code=ACP_TURN_FAILED
2026-04-27T14:58:43 [ws] res ✗ agent errorCode=UNAVAILABLE errorMessage=AcpRuntimeError: Internal error code=ACP_TURN_FAILED
2026-04-27T15:08:16 [ws] res ✗ agent errorCode=UNAVAILABLE errorMessage=AcpRuntimeError: Internal error code=ACP_TURN_FAILED
2026-04-27T15:32:27 [ws] res ✗ agent errorCode=UNAVAILABLE errorMessage=AcpRuntimeError: Internal error code=ACP_TURN_FAILED


WS handshake failure (from /tmp/openclaw/openclaw-2026-04-27.log, structured fields):


{
  "cause": "invalid-handshake",
  "handshake": "failed",
  "lastFrameType": "req",
  "lastFrameMethod": "sessions.list",
  "handshakeError": "invalid handshake: first request must be connect",
  "host": "127.0.0.1:18789"
}



[ERROR] gateway connect failed: Error: gateway closed (1008): invalid handshake: first request must be connect
[WARN] acp startup identity reconcile (renderer=v1): checked=22 resolved=0 failed=22


openclaw doctor (relevant):


Model auth: anthropic:claude-cli expiring (8h)
Claude CLI: Binary ~/.local/bin/claude, workspace writable
Gateway service: PATH includes nvm, Node from version manager
Plugins: Loaded 12, Errors 0


openclaw gateway status:


Service: LaunchAgent (loaded)
Runtime: running, state active
Connectivity probe: ok
Listening: 127.0.0.1:18789


Standalone ACP agent test (works fine outside OpenClaw):


$ echo '{"jsonrpc":"2.0","method":"initialize","id":1,"params":{"protocolVersion":1,...}}' | node claude-agent-acp/dist/index.js
→ {"jsonrpc":"2.0","id":1,"result":{"protocolVersion":1,"agentInfo":{"name":"@agentclientprotocol/claude-agent-acp","version":"0.26.0"}}}
```

### Impact and severity

Affected: All ACP agent spawns (Claude Code, Codex) on this installation
Severity: High — blocks all ACP-based coding agent workflows (Boris/Vlad delegation)
Frequency: 100% — every ACP spawn fails, across 6+ gateway restarts and multiple debugging attempts
Consequence: Cannot use ACP runtime for coding agents; forced to fall back to non-ACP subagent spawns which work but use a different execution model

### Additional information

Last known good: 2026.4.24 (bundled acpx v0.5.3) — ACP working normally on 2026-04-26
First known bad: 2026.4.25 (bundled acpx v0.6.1) — ACP broken on first boot 2026-04-27

Ruled out during debugging:
- Auth: Claude CLI re-authed, fresh token (8h), same failure
- Session state: Cleaned 585 stale session files + 273 orphaned registry entries + 330 transcript files
- Concurrent limit: Bumped maxConcurrentSessions from 40 to 100
- ACP agents: Both claude-agent-acp and codex-acp work correctly standalone

Workaround: Use runtime="subagent" instead of runtime="acp" for agent spawns.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ Docs: https://docs.openclaw.ai
 
 - Plugins/media: auto-enable provider plugins referenced by `agents.defaults.imageGenerationModel`, `videoGenerationModel`, and `musicGenerationModel` primary/fallback refs, so configured Google and MiniMax media providers do not stay disabled behind a restrictive plugin allowlist. Thanks @vincentkoc.
 - Memory-core/dreaming: retry managed dreaming cron registration after startup when the cron service is not reachable yet, so the scheduled Memory Dreaming Promotion sweep recovers without waiting for heartbeat traffic. Fixes #72841. Thanks @amknight.
+- Acpx/runtime: validate the runtime session mode at the `AcpxRuntime.ensureSession` wrapper boundary so callers that pass anything other than `persistent` or `oneshot` get a clear `ACP_INVALID_RUNTIME_OPTION` error instead of silently round-tripping through the encoded handle as a default `persistent` mode and later throwing `SessionResumeRequiredError`. Investigation context: #73071. (#73548) Thanks @amknight.
 
 ## 2026.4.27
 

--- a/extensions/acpx/src/runtime.test.ts
+++ b/extensions/acpx/src/runtime.test.ts
@@ -1,5 +1,5 @@
 import { beforeEach, describe, expect, it, vi } from "vitest";
-import type { AcpRuntime } from "../runtime-api.js";
+import { AcpRuntimeError, type AcpRuntime } from "../runtime-api.js";
 import { AcpxRuntime, __testing } from "./runtime.js";
 
 type TestSessionStore = {
@@ -83,6 +83,43 @@ function makeRuntime(
 describe("AcpxRuntime fresh reset wrapper", () => {
   beforeEach(() => {
     vi.restoreAllMocks();
+  });
+
+  it("rejects unsupported runtime session modes with a clear AcpRuntimeError (issue #73071)", async () => {
+    const baseStore: TestSessionStore = {
+      load: vi.fn(async () => undefined),
+      save: vi.fn(async () => {}),
+    };
+    const { runtime, delegate } = makeRuntime(baseStore);
+    const ensureSpy = vi.spyOn(delegate, "ensureSession").mockResolvedValue({
+      sessionKey: "agent:claude:acp:test",
+      backend: "acpx",
+      runtimeSessionName: "claude",
+    });
+
+    for (const badMode of ["run", "session", "", undefined, null, 0]) {
+      await expect(
+        runtime.ensureSession({
+          sessionKey: "agent:claude:acp:test",
+          agent: "claude",
+          mode: badMode as never,
+        }),
+      ).rejects.toMatchObject({
+        name: "AcpRuntimeError",
+        code: "ACP_INVALID_RUNTIME_OPTION",
+        message: expect.stringContaining("Unsupported ACP runtime session mode"),
+      });
+    }
+
+    expect(ensureSpy).not.toHaveBeenCalled();
+  });
+
+  it("exposes assertSupportedRuntimeSessionMode as a typed guard", () => {
+    expect(() => __testing.assertSupportedRuntimeSessionMode("persistent")).not.toThrow();
+    expect(() => __testing.assertSupportedRuntimeSessionMode("oneshot")).not.toThrow();
+    expect(() => __testing.assertSupportedRuntimeSessionMode("run" as never)).toThrow(
+      AcpRuntimeError,
+    );
   });
 
   it("normalizes OpenClaw Codex model ids for ACP startup", async () => {

--- a/extensions/acpx/src/runtime.ts
+++ b/extensions/acpx/src/runtime.ts
@@ -231,6 +231,25 @@ function failUnsupportedCodexAcpModel(rawModel: string, detail?: string): never 
   );
 }
 
+// acpx's `decodeAcpxRuntimeHandleState` only accepts `persistent` and `oneshot`; any other
+// value silently round-trips through the encoded handle as `persistent` and later throws
+// `SessionResumeRequiredError` on agent restart. Fail fast at this boundary instead.
+// See openclaw/openclaw#73071.
+const SUPPORTED_RUNTIME_SESSION_MODES = new Set(["persistent", "oneshot"] as const);
+
+function assertSupportedRuntimeSessionMode(
+  mode: unknown,
+): asserts mode is "persistent" | "oneshot" {
+  if (typeof mode === "string" && SUPPORTED_RUNTIME_SESSION_MODES.has(mode as never)) {
+    return;
+  }
+  const supported = Array.from(SUPPORTED_RUNTIME_SESSION_MODES).join(", ");
+  throw new AcpRuntimeError(
+    "ACP_INVALID_RUNTIME_OPTION",
+    `Unsupported ACP runtime session mode ${JSON.stringify(mode)}. Expected one of: ${supported}.`,
+  );
+}
+
 function failUnsupportedCodexAcpThinking(rawThinking: string): never {
   throw new AcpRuntimeError(
     "ACP_INVALID_RUNTIME_OPTION",
@@ -460,6 +479,7 @@ export class AcpxRuntime implements AcpRuntime {
   async ensureSession(
     input: Parameters<AcpRuntime["ensureSession"]>[0],
   ): Promise<AcpRuntimeHandle> {
+    assertSupportedRuntimeSessionMode(input.mode);
     const command = resolveAgentCommandForName({
       agentName: input.agent,
       agentRegistry: this.agentRegistry,
@@ -584,6 +604,7 @@ export {
 
 export const __testing = {
   appendCodexAcpConfigOverrides,
+  assertSupportedRuntimeSessionMode,
   codexAcpSessionModelId,
   isCodexAcpCommand,
   normalizeCodexAcpModelOverride,

